hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManagerV2Impl.java

```diff
@@ -0,0 +1,637 @@
+import org.slf4j.LoggerFactory;
+
+import javax.management.ObjectName;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+/**
+ * SCM Pipeline Manager implementation.
+ * All the write operations for pipelines must come via PipelineManager.
+ * It synchronises all write and read operations via a ReadWriteLock.
+ */
+public class PipelineManagerV2Impl implements PipelineManager {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SCMPipelineManager.class);
+
+  private final ReadWriteLock lock;
+  private PipelineFactory pipelineFactory;
+  private PipelineStateManagerV2 stateManager;
+  private Scheduler scheduler;
+  private BackgroundPipelineCreator backgroundPipelineCreator;
+  private final NodeManager nodeManager;
+  private final ConfigurationSource conf;
+  // Pipeline Manager MXBean
+  private ObjectName pmInfoBean;
+  private final SCMPipelineMetrics metrics;
+  private long pipelineWaitDefaultTimeout;
+  private final AtomicBoolean isInSafeMode;
+  // Used to track if the safemode pre-checks have completed. This is designed
+  // to prevent pipelines being created until sufficient nodes have registered.
+  private final AtomicBoolean pipelineCreationAllowed;
+
+  public PipelineManagerV2Impl(ConfigurationSource conf,
+                               NodeManager nodeManager,
+                               PipelineStateManagerV2 pipelineStateManager,
+                               PipelineFactory pipelineFactory)
+      throws IOException {
+    this.lock = new ReentrantReadWriteLock();
+    this.pipelineFactory = pipelineFactory;
+    this.stateManager = pipelineStateManager;
+    this.nodeManager = nodeManager;
+    this.conf = conf;
+    this.pmInfoBean = MBeans.register("SCMPipelineManager",
+        "SCMPipelineManagerInfo", this);
+    this.metrics = SCMPipelineMetrics.create();
+    this.pipelineWaitDefaultTimeout = conf.getTimeDuration(
+        HddsConfigKeys.HDDS_PIPELINE_REPORT_INTERVAL,
+        HddsConfigKeys.HDDS_PIPELINE_REPORT_INTERVAL_DEFAULT,
+        TimeUnit.MILLISECONDS);
+    this.isInSafeMode = new AtomicBoolean(conf.getBoolean(
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_ENABLED,
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_ENABLED_DEFAULT));
+    // Pipeline creation is only allowed after the safemode prechecks have
+    // passed, eg sufficient nodes have registered.
+    this.pipelineCreationAllowed = new AtomicBoolean(!this.isInSafeMode.get());
+    initializePipelineState();
+  }
+
+  public static PipelineManager newPipelineManager(
+      ConfigurationSource conf, SCMHAManager scmhaManager,
```

Comment:
NIT: we can just pass RatisServer instead of SCMHAManager here.

Reply:
We may have more stuff in SCMHAManager than just SCMRatisServer for PipelineManager use cases. The idea here is to have a manager interface for SCM HA so that we won't worry about passing more things into PipelineManager like configs or other things. @xiaoyuyao

Reply:
In the future, we have to do `isLeader` check inside `PipelineManager` for which we might need `SCMHAManager` instance.